File: backend/app/services/settings_service.py
"""Company settings service (Module 12)."""
from __future__ import annotations

from typing import Any

from app.models import CompanySettings
from app.repositories import settings_repo
from app.services import audit_service


def get_settings() -> CompanySettings:
    return settings_repo.get_settings()


def update_settings(data: dict[str, Any], actor=None) -> CompanySettings:
    settings = settings_repo.get_settings()
    for field in (
        "company_name",
        "address",
        "phone",
        "email",
        "logo_path",
        "currency",
        "contract_terms",
    ):
        if field in data and data[field] is not None:
            setattr(settings, field, data[field])
    settings_repo.update(settings, {})
    audit_service.record(
        "update_settings",
        user_id=getattr(actor, "id", None),
        username=getattr(actor, "username", None),
        entity="settings",
        entity_id=settings.id,
    )
    return settings

File: backend-go/internal/service/settings_service.go
package service

import "github.com/instaagrammeta/rentacar/backend-go/internal/models"

// SettingsInput holds updatable company settings fields.
type SettingsInput struct {
	CompanyName   *string `json:"company_name"`
	Address       *string `json:"address"`
	Phone         *string `json:"phone"`
	Email         *string `json:"email"`
	LogoPath      *string `json:"logo_path"`
	Currency      *string `json:"currency"`
	ContractTerms *string `json:"contract_terms"`
}

// UpdateSettings updates the singleton company settings row.
func (s *Service) UpdateSettings(in SettingsInput, actor *Actor) (*models.CompanySettings, error) {
	settings, err := s.GetSettings()
	if err != nil {
		return nil, err
	}
	if in.CompanyName != nil {
		settings.CompanyName = *in.CompanyName
	}
	if in.Address != nil {
		settings.Address = in.Address
	}
	if in.Phone != nil {
		settings.Phone = in.Phone
	}
	if in.Email != nil {
		settings.Email = in.Email
	}
	if in.LogoPath != nil {
		settings.LogoPath = in.LogoPath
	}
	if in.Currency != nil {
		settings.Currency = *in.Currency
	}
	if in.ContractTerms != nil {
		settings.ContractTerms = in.ContractTerms
	}
	if err := s.DB.Save(settings).Error; err != nil {
		return nil, err
	}
	s.record(actor, "update_settings", "settings", uintPtr(settings.ID), "")
	return settings, nil
}
